File: Semester 4/Class12_Computer_Contest_Level_3_Notes_and_Homework_Nov_26/lvl3_hw_sol/AQT_directed_graph.java
import java.io.*;
import java.time.Period;
import java.util.*;

public class homework {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int n, m, vis[]; static List<Integer> adj[];
    public static void main(String[] args) throws IOException {
    	n = readInt(); m = readInt(); vis = new int[n+1];
    	adj = new ArrayList[n+1];
    	for(int i=1; i<=n; i++) adj[i] = new ArrayList<>();
    	for(int i=1; i<=m; i++) {
    		int u = readInt(), v = readInt(); adj[v].add(u);
    	}
    	for(int i=n; i>=1; i--) {
    		if(vis[i] == 0) dfs(i, i);
    	}
    	for(int i=n; i>=1; i--) {
    		if(vis[i] > i) { 
    			System.out.println(i + " " + vis[i]); return;
    		}
    	}
    	System.out.println(-1);
    }
    static void dfs(int u, int mx) {
    	vis[u] = mx;
    	for(int v : adj[u]){
    		if(vis[v] == 0) dfs(v, mx);
    	}
    }
    static String next() throws IOException {
        while (st == null || !st.hasMoreTokens())
            st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }

    static long readLong() throws IOException {
        return Long.parseLong(next());
    }

    static int readInt() throws IOException {
        return Integer.parseInt(next());
    }

    static double readDouble() throws IOException {
        return Double.parseDouble(next());
    }

    static char readCharacter() throws IOException {
        return next().charAt(0);
    }

    static String readLine() throws IOException {
        return br.readLine().trim();
    }
}

File: Semester 4/Class12_Computer_Contest_Level_3_Notes_and_Homework_Nov_26/lvl3_hw_sol/AQT_directed_graph.py
import sys
input = sys.stdin.readline

sys.setrecursionlimit(500000)

n, m = map(int, input().split())
adj = [[] for i in range(n+1)]
for i in range(m):
    u, v = map(int, input().split())
    adj[v].append(u)
vis = [0] * (n+1)

def dfs(u, mx):
    vis[u] = mx
    for v in adj[u]:
        if vis[v] == 0:
            dfs(v, mx)

for i in range(n, 0, -1):
    if vis[i] == 0:
        dfs(i, i)
for i in range(n, 0, -1):
    if vis[i] > i:
        print(i, vis[i])
        sys.exit(0)
print(-1)